File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/CloudGossip.kt
// Kotlin mirror of the Phase-3 cloud-gossip / per-service leadership foundation
// in `packages/protocol/src/cloudGossip.ts` and the Swift mirror
// (FlagshipCore/CloudGossip.swift). PURE crypto / canonical-bytes plumbing —
// MUST stay byte-identical to the TS implementation; the pinned cross-platform
// vectors in `CloudGossipVectorTest.kt` lock every byte/hex in.
//
//   1. CGK   — HKDF-SHA256(ikm = umkSeed, salt = empty,
//              info = "flagship.cloud-gossip.v1", 32). One key PER CLOUD (no
//              serverId) — derived the SAME way as ServerKeys.deriveSwk, only
//              the info differs.
//   2. set-leader — owner-IRK-signed preferred-server vote, mirroring the
//              server-decommission envelope conventions.
//   3. gossip — canonical bytes + an HMAC-SHA256 tag keyed by the CGK, plus an
//              AES-256-GCM nonce-prefixed seal/open transport keyed by the CGK.
//   4. clout  — the pure comparator + per-service elector.
//   5. birthDateFromAuthCode — AuthCode.issuedAt is the immutable birth date.

package com.flagshipserver.app.core

import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

object CloudGossip {
    private val rng = SecureRandom()

    // ── 1. CGK ──────────────────────────────────────────────────────────
    private const val INFO_CGK = "flagship.cloud-gossip.v1"

    /** CGK = HKDF-SHA256(ikm = umkSeed, salt = empty,
     *  info = "flagship.cloud-gossip.v1", 32) — per-cloud (no serverId).
     *  Mirrors ServerKeys.deriveSwk's HKDF construction (empty salt). */
    fun deriveCGK(umkSeed: ByteArray): ByteArray {
        require(umkSeed.size == 32) { "UMK seed must be 32 bytes" }
        return hkdfSha256(umkSeed, INFO_CGK.toByteArray(Charsets.UTF_8))
    }

    // ── 2. set-leader vote ──────────────────────────────────────────────
    const val SET_LEADER_TAG = "flagship/set-leader/v1"
    const val SET_LEADER_NONE = "none"

    /** Owner-IRK-signed preferred-server vote canonical bytes (byte-identical
     *  to TS + Swift):
     *   flagship/set-leader/v1|<user>|<preferredStkPubHex>|<issuedAt>|<nonce>
     *  user, preferredStkPubHex, nonce are lowercased; issuedAt is decimal.
     *  preferredStkPubHex == "none" clears the vote. */
    fun setLeaderCanonicalBytes(
        user: String,
        preferredStkPubHex: String,
        issuedAt: Long,
        nonce: String,
    ): ByteArray =
        listOf(
            SET_LEADER_TAG,
            user.lowercase(),
            preferredStkPubHex.lowercase(),
            issuedAt.toString(),
            nonce.lowercase(),
        ).joinToString("|").toByteArray(Charsets.UTF_8)

    // ── 3. gossip announcement ──────────────────────────────────────────
    const val GOSSIP_TAG = "flagship/gossip/v1"
    const val GOSSIP_VOTE_NONE = "none"
    const val GOSSIP_VOTE_DATE_NONE = 0L

    /** Canonical gossip bytes (byte-identical to TS + Swift):
     *   flagship/gossip/v1|<user>|<name>|<birthAuthHex>|<birthDate>|<voteStkHex>|<voteDate>|<services>|<liveness>|<issuedAt>
     *  services = the slugs SORTED + ","-joined (deterministic). name,
     *  birthAuthHex, voteStkHex lowercased; the dates are decimal. */
    fun gossipCanonicalBytes(
        user: String,
        name: String,
        birthAuthHex: String,
        birthDate: Long,
        voteStkHex: String,
        voteDate: Long,
        services: List<String>,
        liveness: String,
        issuedAt: Long,
    ): ByteArray {
        val svc = services.sorted().joinToString(",")
        return listOf(
            GOSSIP_TAG,
            user.lowercase(),
            name.lowercase(),
            birthAuthHex.lowercase(),
            birthDate.toString(),
            voteStkHex.lowercase(),
            voteDate.toString(),
            svc,
            liveness,
            issuedAt.toString(),
        ).joinToString("|").toByteArray(Charsets.UTF_8)
    }

    /** HMAC-SHA256 of the canonical gossip bytes under the CGK, lowercased hex. */
    fun macGossip(canonical: ByteArray, cgk: ByteArray): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(cgk, "HmacSHA256"))
        return HexUtil.encode(mac.doFinal(canonical))
    }

    /** Constant-time check that `mac` (lowercased hex) is the CGK-HMAC of the
     *  canonical bytes. Never throws. */
    fun verifyGossipMac(canonical: ByteArray, mac: String, cgk: ByteArray): Boolean {
        return try {
            val expected = macGossip(canonical, cgk)
            val got = mac.lowercase()
            if (expected.length != got.length) return false
            var diff = 0
            for (i in expected.indices) diff = diff or (expected[i].code xor got[i].code)
            diff == 0
        } catch (_: Throwable) {
            false
        }
    }

    /** AES-256-GCM transport seal keyed by the CGK. Wire layout (nonce-prefixed):
     *   [nonce: 12 B][ciphertext + GCM tag: var] */
    fun sealGossip(plaintext: ByteArray, cgk: ByteArray): ByteArray {
        val nonce = ByteArray(12).also(rng::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(cgk, "AES"), GCMParameterSpec(128, nonce))
        val ctWithTag = cipher.doFinal(plaintext) // ciphertext ‖ 16-byte GCM tag
        val out = ByteArray(nonce.size + ctWithTag.size)
        System.arraycopy(nonce, 0, out, 0, nonce.size)
        System.arraycopy(ctWithTag, 0, out, nonce.size, ctWithTag.size)
        return out
    }

    /** Open a sealGossip blob with the CGK. Throws AEADBadTagException on a
     *  wrong key / corrupted tag. */
    fun openGossip(blob: ByteArray, cgk: ByteArray): ByteArray {
        require(blob.size >= 12 + 16) { "sealed gossip blob too short" }
        val nonce = blob.copyOfRange(0, 12)
        val ctWithTag = blob.copyOfRange(12, blob.size)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(cgk, "AES"), GCMParameterSpec(128, nonce))
        return cipher.doFinal(ctWithTag)
    }

    // ── 4. clout ranking ────────────────────────────────────────────────
    data class CloutMember(
        val id: String,
        val domain: String,
        val birthDate: Long,
        /** The owner's set-leader vote issuedAt (ms), or null when not voted. */
        val voteIssuedAt: Long?,
        /** "live" | "unreachable" | "never". */
        val liveness: String,
        val services: List<String>,
    )

    /** The raw clout comparator. Returns true when `a` STRICTLY outranks `b`
     *  (i.e. `a` should lead over `b`):
     *    1. higher voteIssuedAt wins (null treated as the lowest);
     *    2. tie → lower birthDate (oldest birth certificate) wins;
     *    3. tie → lower domain lexicographically. */
    fun cloutLess(a: CloutMember, b: CloutMember): Boolean {
        val av = a.voteIssuedAt
        val bv = b.voteIssuedAt
        if (av != bv) {
            return when {
                av != null && bv != null -> av > bv
                av != null && bv == null -> true   // a voted, b not → a leads
                else -> false                       // a not voted, b voted → b leads
            }
        }
        if (a.birthDate != b.birthDate) return a.birthDate < b.birthDate
        return a.domain < b.domain
    }

    /** Elect the leader for one service: among the `live` members that run
     *  serviceSlug, the highest-clout one. null when no live runner exists. */
    fun electLeadForService(members: List<CloutMember>, serviceSlug: String): CloutMember? {
        val eligible = members.filter { it.liveness == "live" && it.services.contains(serviceSlug) }
        if (eligible.isEmpty()) return null
        var lead = eligible[0]
        for (i in 1 until eligible.size) {
            if (cloutLess(eligible[i], lead)) lead = eligible[i]
        }
        return lead
    }

    // ── 5. birth date ───────────────────────────────────────────────────
    /** The immutable birth date: AuthCode.issuedAt (ms). The create-time,
     *  owner-IRK-signed auth code is signed once, so issuedAt is a stable,
     *  unforgeable per-pod birth instant. */
    fun birthDateFromAuthCode(issuedAt: Long): Long = issuedAt

    // ── shared HKDF (mirrors ServerKeys.hkdfSha256) ─────────────────────
    private fun hkdfSha256(ikm: ByteArray, info: ByteArray, length: Int = 32): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(ByteArray(32), "HmacSHA256"))
        val prk = mac.doFinal(ikm)
        mac.init(SecretKeySpec(prk, "HmacSHA256"))
        val out = ByteArray(length)
        var t = ByteArray(0)
        var counter = 1
        var written = 0
        while (written < length) {
            mac.reset()
            mac.update(t)
            mac.update(info)
            mac.update(counter.toByte())
            t = mac.doFinal()
            val n = minOf(t.size, length - written)
            System.arraycopy(t, 0, out, written, n)
            written += n
            counter++
        }
        return out
    }
}

File: apps/mobile/shared/Sources/FlagshipCore/CloudGossip.swift
import Foundation
import CryptoKit

/// Swift mirror of the Phase-3 cloud-gossip / per-service leadership foundation
/// in `packages/protocol/src/cloudGossip.ts`. PURE crypto / canonical-bytes
/// plumbing — must stay byte-identical to the TS implementation and the Kotlin
/// mirror (`CloudGossip.kt`); the pinned cross-platform vectors in
/// `CloudGossipCanonicalTests.swift` lock every byte/hex in.
///
///   1. CGK   — `HKDF-SHA256(ikm = umkSeed, salt = empty, info =
///              "flagship.cloud-gossip.v1", 32)`. One key PER CLOUD (no
///              serverId) — derived the SAME way as `ServerKeys.deriveSwk`,
///              only the info differs.
///   2. set-leader — owner-IRK-signed preferred-server vote, mirroring the
///              server-decommission envelope conventions.
///   3. gossip — canonical bytes + an HMAC-SHA256 tag keyed by the CGK, plus an
///              AES-256-GCM nonce-prefixed seal/open transport keyed by the CGK.
///   4. clout  — the pure comparator + per-service elector.
///   5. birthDateFromAuthCode — AuthCode.issuedAt is the immutable birth date.

public enum CloudGossip {

    // MARK: 1. CGK

    private static let infoCGK = "flagship.cloud-gossip.v1"

    /// `CGK = HKDF-SHA256(ikm = umkSeed, salt = empty,
    /// info = "flagship.cloud-gossip.v1", 32)` — per-cloud (no serverId).
    /// Mirrors `ServerKeys.deriveSwk`'s HKDF construction (empty salt).
    public static func deriveCGK(umkSeed: Data) -> Data? {
        guard umkSeed.count == 32 else { return nil }
        let key = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: umkSeed),
            salt: Data(),
            info: Data(infoCGK.utf8),
            outputByteCount: 32
        )
        return key.withUnsafeBytes { Data($0) }
    }

    // MARK: 2. set-leader vote

    public static let setLeaderTag = "flagship/set-leader/v1"
    /// Sentinel `preferredStkPubHex` that CLEARS the vote.
    public static let setLeaderNone = "none"

    /// Owner-IRK-signed preferred-server vote.
    ///
    /// Canonical bytes (byte-identical to TS + Kotlin):
    ///   flagship/set-leader/v1|<user>|<preferredStkPubHex>|<issuedAt>|<nonce>
    /// `user`, `preferredStkPubHex`, and `nonce` are lowercased; `issuedAt` is
    /// the plain decimal number. `preferredStkPubHex == "none"` clears the vote.
    public struct SetLeaderVote: Equatable, Sendable {
        public let user: String
        public let preferredStkPubHex: String
        public let issuedAt: Int64
        public let nonce: String

        public init(user: String, preferredStkPubHex: String, issuedAt: Int64, nonce: String) {
            self.user = user
            self.preferredStkPubHex = preferredStkPubHex
            self.issuedAt = issuedAt
            self.nonce = nonce
        }

        public func canonicalBytes() -> Data {
            Data(
                [
                    CloudGossip.setLeaderTag,
                    user.lowercased(),
                    preferredStkPubHex.lowercased(),
                    String(issuedAt),
                    nonce.lowercased(),
                ].joined(separator: "|").utf8
            )
        }

        public func sign(with irk: Curve25519.Signing.PrivateKey) throws -> Data {
            try irk.signature(for: canonicalBytes())
        }

        public func verify(_ signature: Data, with irkPub: Curve25519.Signing.PublicKey) -> Bool {
            irkPub.isValidSignature(signature, for: canonicalBytes())
        }
    }

    // MARK: 3. gossip announcement

    public static let gossipTag = "flagship/gossip/v1"
    public static let gossipVoteNone = "none"
    public static let gossipVoteDateNone: Int64 = 0

    /// One pod's per-tick gossip frame.
    ///
    /// Canonical bytes (byte-identical to TS + Kotlin):
    ///   flagship/gossip/v1|<user>|<name>|<birthAuthHex>|<birthDate>|<voteStkHex>|<voteDate>|<services>|<liveness>|<issuedAt>
    /// `services` = the slugs SORTED + `,`-joined (deterministic). `name`,
    /// `birthAuthHex`, `voteStkHex` lowercased; the dates are plain decimals.
    public struct Announcement: Equatable, Sendable {
        public let user: String
        public let name: String
        public let birthAuthHex: String
        public let birthDate: Int64
        public let voteStkHex: String
        public let voteDate: Int64
        public let services: [String]
        /// "live" | "unreachable" | "never".
        public let liveness: String
        public let issuedAt: Int64

        public init(
            user: String, name: String, birthAuthHex: String, birthDate: Int64,
            voteStkHex: String, voteDate: Int64, services: [String],
            liveness: String, issuedAt: Int64
        ) {
            self.user = user
            self.name = name
            self.birthAuthHex = birthAuthHex
            self.birthDate = birthDate
            self.voteStkHex = voteStkHex
            self.voteDate = voteDate
            self.services = services
            self.liveness = liveness
            self.issuedAt = issuedAt
        }

        public func canonicalBytes() -> Data {
            let svc = services.sorted().joined(separator: ",")
            return Data(
                [
                    CloudGossip.gossipTag,
                    user.lowercased(),
                    name.lowercased(),
                    birthAuthHex.lowercased(),
                    String(birthDate),
                    voteStkHex.lowercased(),
                    String(voteDate),
                    svc,
                    liveness,
                    String(issuedAt),
                ].joined(separator: "|").utf8
            )
        }
    }

    /// HMAC-SHA256 of the canonical gossip bytes under the CGK, lowercased hex.
    public static func macGossip(_ a: Announcement, cgk: Data) -> String {
        let tag = HMAC<SHA256>.authenticationCode(
            for: a.canonicalBytes(),
            using: SymmetricKey(data: cgk)
        )
        return Data(tag).map { String(format: "%02x", $0) }.joined()
    }

    /// Constant-time check that `mac` (lowercased hex) is the CGK-HMAC of `a`.
    /// Never throws.
    public static func verifyGossipMac(_ a: Announcement, mac: String, cgk: Data) -> Bool {
        let expected = macGossip(a, cgk: cgk)
        let got = mac.lowercased()
        guard expected.utf8.count == got.utf8.count else { return false }
        var diff: UInt8 = 0
        for (x, y) in zip(expected.utf8, got.utf8) { diff |= x ^ y }
        return diff == 0
    }

    /// AES-256-GCM transport seal keyed by the CGK. Wire layout (nonce-prefixed):
    ///   [nonce: 12 B][ciphertext + GCM tag: var]
    public static func sealGossip(_ plaintext: Data, cgk: Data) throws -> Data {
        let nonceBytes = Data((0..<12).map { _ in UInt8.random(in: 0...255) })
        let nonce = try AES.GCM.Nonce(data: nonceBytes)
        let sealed = try AES.GCM.seal(plaintext, using: SymmetricKey(data: cgk), nonce: nonce)
        // nonce || ciphertext || tag (SealedBox.combined IS nonce||ct||tag).
        return nonceBytes + sealed.ciphertext + sealed.tag
    }

    /// Open a `sealGossip` blob with the CGK. Throws on a bad tag/length.
    public static func openGossip(_ blob: Data, cgk: Data) throws -> Data {
        guard blob.count >= 12 + 16 else {
            throw CryptoKitError.incorrectParameterSize
        }
        let nonce = try AES.GCM.Nonce(data: blob.prefix(12))
        let rest = blob.suffix(from: blob.startIndex + 12)
        let ct = rest.prefix(rest.count - 16)
        let tag = rest.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: SymmetricKey(data: cgk))
    }

    // MARK: 4. clout ranking

    public struct CloutMember: Equatable, Sendable {
        public let id: String
        public let domain: String
        public let birthDate: Int64
        /// The owner's set-leader vote issuedAt (ms), or nil when not voted.
        public let voteIssuedAt: Int64?
        /// "live" | "unreachable" | "never".
        public let liveness: String
        public let services: [String]

        public init(
            id: String, domain: String, birthDate: Int64,
            voteIssuedAt: Int64?, liveness: String, services: [String]
        ) {
            self.id = id
            self.domain = domain
            self.birthDate = birthDate
            self.voteIssuedAt = voteIssuedAt
            self.liveness = liveness
            self.services = services
        }
    }

    /// The raw clout comparator. Returns true when `a` STRICTLY outranks `b`
    /// (i.e. `a` should lead over `b`):
    ///   1. higher `voteIssuedAt` wins (nil treated as the lowest);
    ///   2. tie → lower `birthDate` (oldest birth certificate) wins;
    ///   3. tie → lower `domain` lexicographically.
    public static func cloutLess(_ a: CloutMember, _ b: CloutMember) -> Bool {
        let av = a.voteIssuedAt
        let bv = b.voteIssuedAt
        if av != bv {
            // nil is the lowest; a higher vote is "less" (sorts first / leads).
            switch (av, bv) {
            case let (.some(x), .some(y)): return x > y
            case (.some, .none): return true   // a voted, b not → a leads
            case (.none, .some): return false
            case (.none, .none): break
            }
        }
        if a.birthDate != b.birthDate { return a.birthDate < b.birthDate }
        return a.domain < b.domain
    }

    /// Elect the leader for one service: among the `live` members that run
    /// `serviceSlug`, the highest-clout one. nil when no live runner exists.
    public static func electLeadForService(
        _ members: [CloutMember], serviceSlug: String
    ) -> CloutMember? {
        let eligible = members.filter {
            $0.liveness == "live" && $0.services.contains(serviceSlug)
        }
        guard var lead = eligible.first else { return nil }
        for m in eligible.dropFirst() where cloutLess(m, lead) { lead = m }
        return lead
    }

    // MARK: 5. birth date

    /// The immutable birth date: `AuthCode.issuedAt` (ms). The create-time,
    /// owner-IRK-signed auth code is signed once, so `issuedAt` is a stable,
    /// unforgeable per-pod birth instant.
    public static func birthDateFromAuthCode(issuedAt: Int64) -> Int64 { issuedAt }
}
